src/types/Typechecker/Typechecker.hs

```diff
@@ -1722,33 +1722,43 @@ instance Checkable Expr where
                                         ,stop = eStop
                                         ,step = eStep}
 
-    --  E |- rng : Range
-    --  E, x : int |- e : ty
-    -- --------------------------
-    --  E |- for x <- rng e : ty
-
-    --  E |- arr : [ty]
-    --  E, x : int |- e : ty
-    -- --------------------------
-    --  E |- for x <- arr e : ty
-    doTypecheck for@(For {name, step, src, body}) =
-        do stepTyped <- doTypecheck step
-           srcTyped  <- doTypecheck src
-           let srcType = AST.getType srcTyped
-
-           unless (isArrayType srcType || isRangeType srcType) $
-             pushError src $ NonIterableError srcType
-
-           let elementType = if isRangeType srcType
-                             then intType
-                             else getResultType srcType
-           bodyTyped <- typecheckBody elementType body
-           return $ setType unitType for{step = stepTyped
-                                        ,src  = srcTyped
-                                        ,body = bodyTyped}
-        where
-          addIteratorVariable ty = extendEnvironmentImmutable [(name, ty)]
-          typecheckBody ty = local (addIteratorVariable ty) . typecheck
+
+    -- JOY for-comprehension
+    -- returnType is unittye, is changed during the second typechecking when it has been desugaraed into a method call
```

Comment:
- too long comment. break down into multiple lines if bigger than 80 char threshold
- remove name
- if you can, try to add the typing rules from the comment of the previous for-comprehension implementation